File: daq/slc/tools/javalib/b2dqm/src/b2dqm/core/TimedGraph1.java
package b2dqm.core;

import java.util.Date;

import b2daq.core.Reader;
import b2daq.core.Writer;


public abstract class TimedGraph1 extends Histo {
	
	protected int _iter;
	protected LongArray _time_v;
	
	public TimedGraph1() {
		this("", "", 1, 0, 1);
	}

	public TimedGraph1(String name, String title, int nbinx, double xmin, double xmax) {
		super(name, title);
		_axis_x.setRange(nbinx, xmin, xmax);
		_time_v = new LongArray(nbinx);
	}

	public void reset() {
		_iter = 0;
		_data_v.setAll(0);
		_time_v.setAll(0);
		getAxisY().setMax(1);
	}

	public void fixMaximum(double data, boolean fix) {
		_axis_y.setMax(data);
		_axis_y.fixMax(fix);
	}

	public void fixMinimum(double data, boolean fix) {
		_axis_y.setMin(data);
		_axis_y.fixMin(fix);
	}

	public int getDim() { return 1; }
	public double getMaximum() { return _axis_y.getMax(); }
	public double getMinimum() { return _axis_y.getMin(); }
	public void setMaximum(double data) {
		if ( !_axis_y.isFixMax() ) _axis_y.setMax(data);
	}

	public void setMinimum(double data) { 
		if ( !_axis_y.isFixMin() ) _axis_y.setMin(data);
	}

	public double getPoint(int n) {
		return _data_v.get(n);
	}
	
	public void setPoint(int n, double data) { 
		_data_v.set(n, data);
	}
	
	public long getTime(int n) {
		return (long)_time_v.get(n);
	}
	
	public void setTime(int n, long time) { 
		_time_v.set(n, time);
	}
	
    public void addPoint(long time, double data) {
    	_iter++;
    	if ( _iter == _data_v.length() || 
    		 _iter == _time_v.length() ) {
    		_iter = 0;
    	}
    	setPoint(_iter, data);
    	setTime(_iter, time);
    }

    public void addPoint(double data) {
    	addPoint(new Date().getTime(), data);
    }
    	
    public double getLatestPoint() {
    	return _data_v.get(_iter);
    }
    
    public long getLatestTime() {
    	return (long)_time_v.get(_iter);
    }

    public void setMaxAndMin(){
		if ( ! _axis_y.isFixMax() || ! _axis_y.isFixMin()) {
			double data, min = getMinimum(), max = getMaximum(); 
			for ( int nx = 0;  nx < getAxisX().getNbins(); nx++) {
				data = getPoint(nx);
				if ( data > max ) max = data;
				if ( data < min ) min = data;
			}
			setMaximum((max+min)/2. + (max-min)/2.*1.05);
			setMinimum((max+min)/2. - (max-min)/2.*1.05);
		}
	}

	public String toString() {
		StringBuffer script = new StringBuffer();
		script.append("      {\n" + "        'datatype':'" + getDataType()
				+ "',\n" + "        'name':'" + getName() + "',\n"
				+ "        'title':'" + getTitle() + "',\n"
				+ "        'tabid':" + (int) getTabId() + ",\n"
				+ "        'positionid':" + (int) getPositionId() + ",\n"
				+ "        'axisx':" + getAxisX().toString() + ",\n"
				+ "        'axisy':" + getAxisY().toString() + ",\n"
				+ "        'axisz':" + getAxisZ().toString() + ",\n"
				+ "        'iter':" + _iter + "\n"  
				+ "        'time':" + _time_v.toString() + "\n"
				+ "        'data':" + _data_v.toString() + "\n" + "      }");
		return script.toString();
	}

	protected String getXML() {
		String str = "title=\"" + getTitle() + 
		"\" data=\"" + _data_v.toString() + 
		"\" iter=\"" + _iter + 
		"\" >\n";
		str += _axis_x.getXML("axis-x");
		str += _axis_y.getXML("axis-y");
		return str;
	}

	@Override
	public void readObject(Reader reader) throws Exception {
		readConfig(reader);
		_iter = reader.readInt() - 1;
		getData().readObject(reader);
		_time_v.readObject(reader);
	}

	@Override
	public void readContents(Reader reader) throws Exception {
		_iter = reader.readInt();
		_time_v.readObject(reader);
		getData().readObject(reader);
	}

	@Override
	public void readConfig(Reader reader) throws Exception {
		setTabId(reader.readChar());
		setPositionId(reader.readChar());
		setName(reader.readString());
		setTitle(reader.readString());
		setAxisX(readAxis(reader));
		getAxisY().setTitle(reader.readString());
		_data_v.resize(_axis_x.getNbins());
		_time_v.resize(_axis_x.getNbins());
	}

	@Override
	public void writeObject(Writer writer) throws Exception {
		writeConfig(writer);
		writer.writeInt(_iter);
		_data_v.writeObject(writer);
		_time_v.writeObject(writer);
	}

	@Override
	public void writeContents(Writer writer) throws Exception {
		writer.writeInt(_iter);
		_time_v.writeObject(writer);
		_data_v.writeObject(writer);
	}

	@Override
	public void writeConfig(Writer writer) throws Exception {
		super.writeConfig(writer);
		writeAxis(writer, getAxisX());
		writer.writeString(getAxisY().getTitle());
	}

	public int getIter() {
		return _iter;
	}

}

File: daq/slc/dqm/src/TimedGraph1.cc
#include "daq/slc/dqm/TimedGraph1.h"

#include "daq/slc/system/Time.h"

#include <sstream>

using namespace Belle2;

TimedGraph1::TimedGraph1()
  : Histo("", ""), m_iter(0), m_time_v(1)
{
  m_axis_x = Axis(1, 0, 1, "");
}

TimedGraph1::TimedGraph1(const std::string& name,
                         const std::string& title,
                         int nbinx, double xmin, double xmax)
  : Histo(name, title), m_iter(0), m_time_v(nbinx)
{
  m_axis_x = Axis(nbinx, xmin, xmax, m_axis_x.getTitle());
}

TimedGraph1::~TimedGraph1() throw()
{
}

void TimedGraph1::reset() throw()
{
  m_iter = 0;
  m_data->setAll(0);
  m_axis_y.setMax(1);
}

void TimedGraph1::setMaximum(double data) throw()
{
  m_axis_y.setMax(data);
}

void TimedGraph1::setMinimum(double data) throw()
{
  m_axis_y.setMin(data);
}

void TimedGraph1::fixMaximum(double data) throw()
{
  m_axis_y.fixMax(true);
  m_axis_y.setMax(data);
}

void TimedGraph1::fixMinimum(double data) throw()
{
  m_axis_y.fixMin(true);
  m_axis_y.setMin(data);
}

double TimedGraph1::getPoint(int n) const throw()
{
  return m_data->get(n);
}

void TimedGraph1::setPoint(int n, double data) throw()
{
  m_data->set(n, data);
}

long TimedGraph1::getTime(int n) const throw()
{
  return (long)m_time_v.get(n);
}

void TimedGraph1::setTime(int n, long long time) throw()
{
  m_time_v.set(n, time);
}

void TimedGraph1::addPoint(long long time, double data) throw()
{
  m_iter++;
  if (m_iter == (int)m_data->size() ||
      m_iter == (int)m_time_v.size()) {
    m_iter = 0;
  }
  setPoint(m_iter, data);
  setTime(m_iter, time);
}

void TimedGraph1::addPoint(double data) throw()
{
  addPoint(Belle2::Time().getSecond(), data);
}

double TimedGraph1::getLatestPoint() const throw()
{
  return m_data->get(m_iter);
  if (m_iter > 0) return m_data->get(m_iter - 1);
  else return m_data->get(m_data->size() - 1);
}

long TimedGraph1::getLatestTime() const throw()
{
  return m_time_v.get(m_iter);
  if (m_iter > 0) return m_time_v.get(m_iter - 1);
  else return m_time_v.get(m_time_v.size() - 1);
}

void TimedGraph1::readObject(Belle2::Reader& reader)
throw(Belle2::IOException)
{
  readConfig(reader);
  m_iter = reader.readInt();
  m_data->readObject(reader);
  m_time_v.readObject(reader);
}

void TimedGraph1::readUpdate(Belle2::Reader& reader)
throw(Belle2::IOException)
{
  long long t = reader.readLong();
  double d = m_data->readElement(reader);
  addPoint(t, d);
}

void TimedGraph1::readConfig(Belle2::Reader& reader)
throw(Belle2::IOException)
{
  Histo::readConfig(reader);
  m_axis_x.readObject(reader);
  m_axis_y.setTitle(reader.readString());
  m_data->resize(m_axis_x.getNbins());
  m_time_v.resize(m_axis_x.getNbins());
}

void TimedGraph1::readContents(Belle2::Reader& reader)
throw(Belle2::IOException)
{
  m_iter = reader.readInt();
  m_time_v.readObject(reader);
  m_data->readObject(reader);
}

void TimedGraph1::writeObject(Belle2::Writer& writer)
const throw(Belle2::IOException)
{
  writeConfig(writer);
  writer.writeInt(m_iter);
  m_data->writeObject(writer);
  m_time_v.writeObject(writer);
}

void TimedGraph1::writeUpdate(Belle2::Writer& writer)
const throw(Belle2::IOException)
{
  writer.writeLong(getLatestTime());
  m_data->writeElement(getLatestPoint(), writer);
}

void TimedGraph1::writeConfig(Belle2::Writer& writer)
const throw(Belle2::IOException)
{
  Histo::writeConfig(writer);
  m_axis_x.writeObject(writer);
  writer.writeString(m_axis_y.getTitle());
}

void TimedGraph1::writeContents(Belle2::Writer& writer)
const throw(Belle2::IOException)
{
  writer.writeInt(m_iter);
  m_time_v.writeObject(writer);
  m_data->writeObject(writer);
}

std::string TimedGraph1::toString() const throw()
{
  std::stringstream script("");
  script << "<histo data-type='" << getDataType() << "' "
         << "name='" << getName() << "' "
         << "title='" << getTitle() << "' "
         << "tab='" << (int) getTabId() << "' "
         << "position='" << (int) getPositionId() << "' " << std::endl
         << "time='" << m_time_v.toString() << "' " << std::endl
         << "data='" << m_data->toString() << "' >" << std::endl
         << "<axis-x nbins='" << getAxisX().getNbins() << "' "
         << "min='" << getAxisX().getMin() << "' "
         << "max='" << getAxisX().getMax() << "' "
         << "title='" << getAxisX().getTitle() << "' />" << std::endl
         << "<axis-y nbins='" << getAxisY().getNbins() << "' "
         << "min='" << getAxisY().getMin() << "' "
         << "max='" << getAxisY().getMax() << "' "
         << "title='" << getAxisY().getTitle() << "' />" << std::endl
         << "</histo>" << std::endl;

  return script.str();
}

